### PR TITLE
Add funder payment reminder button to front end issue, email reminders to backend management command

### LINKED ISSUE #2204: [CODE] Funder Payout Reminders
### User Story

[comment]: # (As a <user type>, I want to <task> so that <goal>.)

As a Gitcoin User, I want to be sure that people are being paid promptly for the time they spend developing on the platform so other great developers trust the platform.

### Why Is this Needed

[comment]: # (Describe the problem and why this task is needed. Provide description of the current state, what you would like to happen, and what actually happen)
*Summary*: Right now, funders sometimes forget to payout bounty hunters after the work is completed. There should be ways to remind / prompt the funder to payout when it is clear that the work has been completed. 

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

### Current Behavior
[comment]: # (Describe what actually happened.)
When work is merged, there are no reminders given to the funder. 

## Solution
[comment]: # (Provide a summary of the solution and a task list on what needs to be fixed.)

### Part I: Gitcoin Flow 
1. Need to look for a scenario where a) the Github issue is closed and b) there is a related PR which the Gitcoiner accomplished which is merged. If these two things are true, move to (2). 
2. A reminder e-mail is sent to the funder if the above two conditions are met for 24 hours **without payment.**
3. After three days, @gitcoinbot pings the thread reminding the funder that this issue is good for payout (one-time reminder). 
4. If it still is not paid out in 3 days, bounty is 'escalated to moderators'. 

If payment is made at any time during the process, the bounty is removed from the above flow and no future steps occur. 

### Part II: Developer 'Payout Reminder' Button 
A simple 'Send Payout Reminder' button on the Issue Details page only available to Gitcoiner who **has submitted work** for a task. 

screen shot 2018-09-12 at 9 56 25 am
  
- [ ] Add the 'Send Payout Reminder' button here, using proper spacing and styling. 
- [ ] When button is clicked, open a modal which says "This button sends a notification to the funder to pay out the bounty. Are you sure the bounty has been completed?" 
- [ ] Ensure only one e-mail can be sent to the funder

### Definition of Done
[comment]: # (Any other information that would be useful, bullets are helpful.)

- [ ] The above flow is tested, implemented, and passes all tests
- [ ] A video walkthrough showing the implementation of both Part II and would be helpful 

### Additional Information
[comment]: # (Any other information that would be useful, content, screenshots, etc.)
- Please allow 3 rounds of reviews with @thelostone-mc @PixelantDesign @SaptakS 
- Pls review before bounty, @PixelantDesign 🙂

### PR DESCRIPTION
**Pull request by oogetyboogety**
@owocki @PixelantDesign @thelostone-mc Please review these changes for merge.
Added funder payment reminder button to javascript, model to db, and REST action to Django

Description
Adding a button for a contributor to remind a funder to pay out a bounty as well as a management commands that runs on an interval specified in crontab. The command posts comments and sends emails on aging BountyFulfillments that have not been accepted yet accroding to the thresholds set out in the Issue

Checklist

- [X] Read and conforms to the contributor guidelines
- [X] Read and conforms to the style guidelines
- [X] Linter status: 100% pass
- [X] Added relevant tests
- [X] Tests pass and code coverage has not decreased
- [X] Changes don't break existing behavior
- [X] Commit message follows commit guidelines
- [X] Code documentation
- [X] Completed: Code review by core team and any requested changes
Affected core subsystem(s)
Refers/Fixes
#2204
Testing and Sign-off
- [X] Ran it through my django setup-everything appears to work as it should.
- [X] Added tests for email and management command functionality

Contributor

- [X] Read and followed the Contributor Guidelines
- [X] Tested all changes locally
- [X] Verified existing functionality
- [X] Ran make test and everything passed!
Reviewer
- [X] Affirm contributor guidelines have been followed and requested changes made
- [X] CI tests and linting pass
- [X] No conflicts (migrations, files, etc)
- [ ] Regression tested against staging
Funder
- [ ] Validated requested changes were made to specification
- [ ] Bounty payout released to the contributor
Screenshots are here:
screenshot from 2019-01-15 01-26-08
screenshot from 2019-02-05 00-26-01
screenshot from 2019-01-15 00-43-00
screenshot from 2019-01-15 00-35-13

### CODE REVIEW

**Comment by octavioamu** [comment]
> * how easy would it be to shift to bootstrap modal
Here is an example 
https://github.com/gitcoinco/web/pull/3592/files

> * add a notification for funder when a reminder is sent to them
Can take a look at the notifications signals file.

**Comment by thelostone-mc** [comment]
@oogetyboogety mind doing what @octavioamu has listed there ?

File: app/retail/emails.py [hunk 1]
```diff
@@ -1092,6 +1099,24 @@ def bounty_feedback(request):
...
+        username (str): The Github username to reference in the email.
+
+    Returns:
+        HttpResponse: The HTML version of the templated HTTP response.
+
+    """
+    bounty = Bounty.objects.filter(fulfillment_submitted_on__isnull=False).first()
+    github_username = request.GET.get('username', '@foo')
+    response_html, _ = render_funder_payout_reminder(username, limit, duration_copy)
+    return HttpResponse(response_html)
```
Comment (thelostone-mc) [inline app/retail/emails.py]: ^ needs to be re-written ! Doesn't work at the moment aka the admin view for the email

**Comment by oogetyboogety** [comment]
@thelostone-mc Alright, this involves some fundamental changes, but I guess I can get it done quickly.

**Comment by oogetyboogety** [comment]
@thelostone-mc @octavioamu I've tried implementing the modal as shown by @octavioamu but I am not getting the click handler to register. I thought I copied what @octavioamu was doing...@thelostone-mc could you please help me figure this out?

**Comment by oogetyboogety** [comment]
@thelostone-mc @octavioamu I've pushed my attempt, so please look over and review and let me know when you have. Thanks for your expedience!

**Comment by owocki** [comment]
note to self: this is a continuation PR of https://github.com/gitcoinco/web/pull/3536

**Comment by oogetyboogety** [comment]
Shot of the new modal
screenshot from 2019-01-28 23-32-57

**Comment by oogetyboogety** [comment]
@thelostone-mc Fixed the merge conflict

File: app/dashboard/templates/bounty/details.html [hunk 2]
```diff
@@ -204,7 +204,7 @@ <h5 class="bounty-heading">{% trans "Funder" %}</h5>
             <span class="font-caption">[[:text]] ( [[:pending_acceptance]] )</span>
           </a>
         [[else color == 'white']]
-          <a class="btn btn-small [[if !enabled]]disabled[[/if]] [[:buttonclass]]" id="btn-white" role="button" target="_blank" href="[[:href]]" target="[[:target]]">
+          <a class="btn btn-small [[if !enabled]]disabled[[/if]] [[:buttonclass]]" id="btn-white" role="button" target="_blank" href="[[:href]]" target="[[:target]]" >
```
Comment (danlipert) [inline app/dashboard/templates/bounty/details.html]: Looks like an extra whitespace snuck in here

File: app/dashboard/utils.py [hunk 3]
```diff
@@ -506,6 +506,18 @@ def get_ordinal_repr(num):
     return f'{num}{suffix}'
 
 
+def record_funder_inaction_on_fulfillment(bounty_fulfillment):
+    payload = {
+        'proflie': bounty_fulfillment.bounty.bounty_owner_profile,
```
Comment (danlipert) [inline app/dashboard/utils.py]: Small typo here: `proflie` -> `profile`

File: app/dashboard/views.py [hunk 4]
```diff
@@ -1100,6 +1099,59 @@ def bounty_details(request, ghuser='', ghrepo='', ghissue=0, stdbounties_id=None
...
+
+    has_fulfilled = bounty.fulfillments.filter(fulfiller_github_username=github_user_data['login']).count()
+    if has_fulfilled == 0:
+        return JsonResponse({
+            'success': False,
+          },
+          status=403)
+
+    #  410 Gone Indicates that the resource requested is no longer available and will not be available again.
+    if bounty.funder_last_messaged_on:
```
Comment (danlipert) [inline app/dashboard/views.py]: Is the idea here that the funder can only be reminded once?

File: app/git/utils.py [hunk 5]
```diff
@@ -573,9 +573,12 @@ def get_interested_actions(github_url, username, email=''):
 
             for pr_action in all_pr_actions:
                 if 'actor' in pr_action:
-                    gh_user = pr_action['actor']['login']
+                    if (pr_action['actor']):
+                        gh_user = pr_action['actor']['login']
```
Comment (danlipert) [inline app/git/utils.py]: What do you think of a oneliner here, like `gh_user = pr_action.get('actor', {}).get('login')`?

File: app/marketing/mails.py [hunk 6]
```diff
@@ -487,6 +487,27 @@ def gdpr_reconsent(email):
     )
 
 
+def funder_payout_reminder(to_email, bounty, github_username, live):
```
Comment (danlipert) [inline app/marketing/mails.py]: Can we add a test for this email to keep our code coverage up? Thanks!

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
+'''
+    Copyright (C) 2017 Gitcoin Core
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: Lets bump the year up here :)

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+logger = logging.getLogger('')
+console = logging.StreamHandler()
+console.setLevel(logging.INFO)
+console.setFormatter(logging.Formatter('%name)s - %(levelname)s - %(message)s'))
+logger.addHandler(console)
+
+
+class Command(BaseCommand):
+
+    help = 'pulls github pull request event sassociated with unaccepted BountyFulfillments'
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: typo "sassociated" here

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    help += ' then notifies the funder by email the issue is open awaiting their review.'
+    help += ' Also escalates bounties that are being held open even with an accepted PR.'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-live', '--live', action='store_true', dest='live', default=False, help='Actually Send the emails'
+        )
+
+    def _get_time_window(self, timestamp, settings):
+        # This is the subset of profile IDs to be saved for PullRequest Eents in the GitHub Events API
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: typo "Eents"

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        hours = 500 if settings.DEBUG else 24
+        start = timezone.now() - timezone.timedelta(hours=hours)
+        deadline = timezone.now() - timezone.timedelta(hours=hours * 3)
+        escalated_deadline = deadline - timezone.timedelta(hours=hours * 3)
+        return hours, start, deadline, escalated_deadline
+
+    def _check_if_bounty_is_closed(self, bounty, bounty_fulfillment):
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: Unused argument here `bounty_fulfillment`

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+from marketing.mails import funder_payout_reminder
+
+logger = logging.getLogger('')
+console = logging.StreamHandler()
+console.setLevel(logging.INFO)
+console.setFormatter(logging.Formatter('%name)s - %(levelname)s - %(message)s'))
+logger.addHandler(console)
+
+
+class Command(BaseCommand):
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: Would definitely love some tests around this management command :+1:

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+                to_email=bounty.bounty_owner_email,
+                bounty=bounty,
+                github_username=bounty_fulfillment.fulfiller_github_username,
+                live=live
+            )
+            if bounty_fulfillment.created_on < deadline:
+                logger.info('Posting github comment')
+                try:
+                    post_issue_comment(
+                        bounty.github_org_name, bounty.gihtub_repo_name, bounty.github_issue_number,
```
Comment (danlipert) [inline app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: typo here "gihtub"

File: app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py (hunk 11 above)
Comment (oogetyboogety) [reply app/marketing/management/commands/sync_pull_request_with_bounty_fulfillments.py]: @danlipert Just added a test to the management command, please review

File: app/dashboard/views.py (hunk 4 above)
Comment (oogetyboogety) [reply app/dashboard/views.py]: @danlipert Exactly!

File: app/git/utils.py (hunk 5 above)
Comment (danlipert) [reply app/git/utils.py]: I think its pretty slick personally, but I can see how it may be a bit unclear whats happening. What did you mean by "nomad"?

**Comment by oogetyboogety** [comment]
Haha i had to look up what happens to the empty dictionary and why it doesn't evaluate to an AttributeError, that's all.

oops typo, i think i meant monad
________________________________

**Comment by danlipert** [comment]
@oogetyboogety ah monad makes more sense :) What's your concern with running the test cases, and why are the named `manual_...`? They seem to run fine in the test suite with `make tests`

**Comment by oogetyboogety** [comment]
They seem to fail in travis when resolving static assets in the template for some reason
________________________________

**Comment by danlipert** [comment]
> They seem to fail in travis when resolving static assets in the template for some reason
@oogetyboogety I see them running in Travis fine - maybe I'm missing something here but it all looks good to me! Can you just rename those tests to remove the `manual_`? I don't want anyone to misattribute any kind of meaning to that prefix in the future.

**Comment by oogetyboogety** [comment]
@danlipert done!

**Comment by danlipert** [comment]
@oogetyboogety nice! thanks for the quick turnaround!

**Comment by oogetyboogety** [comment]
@danlipert Thank _you_!

**Comment by oogetyboogety** [comment]
@danlipert Updated to mock everything like static assets and external API calls in tests. This makes it rely less on state and fail less while still checking the correct behavior. Take a look!

@thelostone-mc This one should be ready for staging.

**Comment by oogetyboogety** [comment]
@danlipert I think we might be good yo merge with your apptoval. Thanks for all your help on this.

**Comment by danlipert** [comment]
@PixelantDesign if this looks good to you lets merge it in 👍 I can't put it on a staging instance yet, but @oogetyboogety has some screenshots here to check out

**Comment by PixelantDesign** [comment]
Hi @oogetyboogety If the functionality is there then that's great!

Would it be possible to clean it up visually and make the fonts and spacing consistent with the modal for express interest? It's okay that this modal has less content, and might naturally be shorter in height.

See this screenshot below for reference:
screen shot 2019-02-15 at 9 34 06 am

**Next Steps Recommendations:**

- [ ] Add a close button in the upper right hand corner
- [ ] Instead of a button that says Yes, could we change language to 'Send Reminder'?
- [ ] Check spacing and location of current button placement.
- [ ] Check horizontal and vertical spacing of elements, and relationships to each other. Refer to Express Interest Modal.
- [ ] Left a sign body copy to be consistent with Express Interest modal.

I'll tip for these changes. Thanks!

**Comment by oogetyboogety** [comment]
@PixelantDesign Addded in the changes you recommended, please let me know what you think as soon as you get a chance :-)

screenshot from 2019-02-18 19-16-22

**Comment by oogetyboogety** [comment]
@PixelantDesign 
Thanks for the review, I've added the changes you recommended. Please review at your earliest convenience!
screenshot from 2019-02-18 21-41-07

Thanks again

**Comment by oogetyboogety** [comment]
@danlipert I think Pixelant said to get with you to follow up

**Comment by danlipert** [comment]
@oogetyboogety sounds good - I'll speak with @PixelantDesign today and then we can get this merged - in the meantime, can you please rebase the PR? thanks!

**Comment by oogetyboogety** [comment]
@danlipert Rebased! Thanks so much!

**Comment by oogetyboogety** [comment]
@danlipert I had to create a new pull request here.
https://github.com/gitcoinco/web/pull/3894
@owocki @danlipert @thelostone-mc @PixelantDesign I think this is the last time I will try to merge changes. if you guys can't catch this one up, I give up, lol.

**Pull request closed without merging**